### DISCUSSION TITLE
[v2] How to open a file in an existing window instead of creating a new window?

### DISCUSSION THREAD

**Discussion by justDeek** [tauri-apps/tauri#10997] [Q&A] [answered]
I am currently working on a text editor app that closes to tray when exiting, meaning the app is still running in the background with the window hidden and when I open a new text file, I want it to be opened in the existing, hidden window, but it instead creates a new window with a second tray icon, both being created with the label 'main'.

I already set up the general logic to achieve this, where I check for an existing webview-window at the start of setup() and if none could be found, it creates a new window.

Here's a simplified version of my current approach:
```rust
fn create_window(app: tauri::AppHandle) {
  tauri::WebviewWindowBuilder::new(&app, "main", tauri::WebviewUrl::default())
    .build()
    .unwrap();
}

pub fn run() {
  let app = tauri::Builder::default()
    .setup(|app| {
      let handle = app.handle();
      
      if let Some(window) = handle.get_webview_window("main") {
        if window.is_visible().unwrap() {
          window.set_focus().unwrap();
        } else {
          window.show().unwrap();
          window.set_focus().unwrap();
        }
      } else {
        create_window(handle.clone());
        //...tray initilization logic...
      }

      Ok(())
    })
    .build(tauri::generate_context!())
    .expect("error while running tauri application");
  
  app.run(|_app_handle, _event| {});
}
```

My current problem is that ```handle.get_webview_window("main")``` doesn't find anything, even when two windows with that label are open. I also tried different window-related functions like ```get_window``` or ```get_webview```, which don't seem to be working on v2. I also tried iterating through the ```webview_windows``` hashmap in run(), which only returns one window labeled 'main', even though two are actively open.

So my main issue is actually retrieving all open windows.
Maybe when calling tauri::Builder it creates a new, independent instance without any shared state.
I was also going through the multiwindow and multiwebview examples/docs, but they only focus on creating the windows/webviews, not retrieving all of them.

I also don't want enforce a [single-instance](https://github.com/tauri-apps/plugins-workspace/tree/v2/plugins/single-instance), because I still want the user to be able to open multiple text files in separate windows for side-by-side comparison, but when the user closes one window to the tray and opens another file,  that file should be opened in the hidden tray window.

I chose this closing-to-tray behavior because without it, opening a new instance of the app whenever opening a file has a noticeable delay with a short flash in the beginning, even though my app is quite minimal. That seems to be a general problem with tauri apps (maybe because of the webview being created and maybe just/most noticeable on windows).

Is there something I'm missing, like a simpler approach, or is my desired behavior not possible?
___
result of ```tauri info```:
[✔] Environment
    - OS: Windows 10.0.22621 X64
    ✔ WebView2: 128.0.2739.79
    ✔ MSVC: Visual Studio Build Tools 2022
    ✔ rustc: 1.80.1 (3f5fd8dd4 2024-08-06)
    ✔ cargo: 1.80.1 (376290515 2024-07-16)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 21.6.2
    - npm: 10.5.0
    - bun: 1.1.18

[-] Packages
    - tauri [RUST]: 2.0.0-rc.10
    - tauri-build [RUST]: 2.0.0-rc.9
    - wry [RUST]: 0.43.1
    - tao [RUST]: 0.30.0
    - @tauri-apps/api [NPM]: 2.0.0-rc.4
[-] App
    - build-type: bundle
    - CSP: default-src 'self' style-src 'self' 'unsafe-inline'; connect-src ipc: http://ipc.localhost
    - frontendDist: ../desktop/build
    - devUrl: http://localhost:3000/
    - framework: Svelte
    - bundler: Vite

**Comment by FabianLars** [comment, marked as answer]
So i assume with opening a text file you're talking about file association (double click on the text file automatically opens it in your app)? If so then you do need something like the single-instance plugin because Windows will always spawn a new **app** instance to open the file, this is just how file assocations work so really no way around it.

Also, just to be sure: app instances != window instances. The single-instance plugin prevents multiple app instances (and therefore multiple tray icons too) but does not care about the amount of windows belonging to that one main instance.

**Reply by justDeek** [reply]
okay, good to know

**Reply by theinfinit** [reply]
> Also, just to be sure: app instances != window instances. The single-instance plugin prevents multiple app instances (and therefore multiple tray icons too) but does not care about the amount of windows belonging to that one main instance.

Brilliant explanation @FabianLars. This is the exact information I was missing from description of Single Instance plugin 😆

Best!
